src/wallet/rpc/util.h

```diff
@@ -40,7 +41,7 @@ const LegacyScriptPubKeyMan& EnsureConstLegacyScriptPubKeyMan(const CWallet& wal
 
 bool GetAvoidReuseFlag(const CWallet& wallet, const UniValue& param);
 bool ParseIncludeWatchonly(const UniValue& include_watchonly, const CWallet& wallet);
-std::string LabelFromValue(const UniValue& value);
+const std::string& LabelFromValue(const UniValue& value LIFETIMEBOUND);
```

Comment:
Guess that you wanted to set `LIFETIMEBOUND` for the return type instead of the function's argument right?.

Still, would return the plain string and not a reference. There is no need to overcomplicate code for address book labels.

Reply:
Thanks, removed the reference.

Reply:
> Guess that you wanted to set LIFETIMEBOUND for the return type instead of the function's argument right?

The return value of `LabelFromValue` was bound to the lifetime of `const UniValue& value` so the attribute was correct I think? See https://clang.llvm.org/docs/AttributeReference.html#lifetimebound

> There is no need to overcomplicate code for address book labels.

I'm honestly not sure why this got reverted. It introduces almost no additional complexity in the code, and callsites can now _choose_ if they wish to have `LabelFromValue` return by value or by reference. Even though wallet labels aren't expected to be huge strings, I don't see the downside of doing it this way? There's no point bikeshedding over this, we won't be able to measure the performance impact, but it just seems like the more proper way of doing things, imo. Up to you, @aureleoules.

Reply:
> I'm honestly not sure why this got reverted. It introduces almost no additional complexity in the code, and callsites can now choose if they wish to have LabelFromValue return by value or by reference. Even though wallet labels aren't expected to be huge strings, I don't see the downside of doing it this way? There's no point bikeshedding over this, we won't be able to measure the performance impact, but it just seems like the more proper way of doing things, imo. Up to you, @aureleoules.

I'm actually not really happy with the revert neither but for another reason.

But first, my point above was to take advantage of the compiler's [rvo](https://stackoverflow.com/questions/12953127/what-are-copy-elision-and-return-value-optimization), which is designed for this kind of stuff.

Now, the reason why I'm not particularly happy with the latest changes is the use of the `auto` keyword alone (which involves a string copy). Instead, should receive the function's output into a const ref string. @aureleoules.

Reply:
Is this ok with you? https://github.com/bitcoin/bitcoin/compare/cce2e872e1a8b7c5b80036f48f1f35467b9328ff..b119bb84250f5b97053314f730aefb8c10b4867a

Reply:
Have spoken with @stickies-v further and he made me notice that the string is not initialized inside `LabelFromValue`, the univalue `get_str` method returns a const reference.. so the compiler's rvo argument does not apply.

So all good for me now, I'm aligned on returning the const ref string from `LabelFromValue` and adding the lifetime bound attribute once more.
just need to re-add the following line:
```c++
const std::string& LabelFromValue(const UniValue& value LIFETIMEBOUND);
```

Reply:
Can be marked as resolved, as furszy mentioned we're all in agreement here and I'm okay with the current implementation 👍

Reply:
Changing this to return a reference doesn't seem very safe -- in that case calling `string& label = LabelFromValue(data["label"])` will mean `label` can become invalid if `data` gets changed (eg, if additional calls to `data.pushKV()` are made, the `data.values` vector may be reallocated, invalidating the reference), which is not something `LIFETIMEBOUND` can check.

As far as I can see we're talking one copy of each fairly short label per RPC call here, which isn't something that needs to be aggressively optimised. So seems much better to me to just keep the implicit copy of the string and have a simple API that minimises the risk of errors creeping in.

Reply:
That's a great point I hadn't considered. Even though the `data["label"]` `UniValue::VSTR` object that we bind to doesn't use a vector to store its string value and would seem safe at first, `data["label"]` returns a reference to the `data.values` vector - so any changes to that parent `UniValue::VOBJ` that we can't control for here are potentially dangerous.

I agree that this could lead to nasty issues and isn't worth the overhead. Sorry for causing the back and forth on this @aureleoules. The `UniValue` type can be a bit tricky to reason about, it seems.

Reply:
Done https://github.com/bitcoin/bitcoin/compare/408e375d15e327b2f5886f395dbc23a4834460e5..632202c5b7bb5fb975f8bde74b6085772b142c5b